### PR TITLE
feat(metrics): add webrtc metrics

### PR DESCRIPTION
**Pull request by mccarthytyler**
# COMPLETES [SPARK-443328](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-443328)

## This pull request addresses

Adding webrtc metrics from wcme.

## by making the following changes

Adding a RtcMetrics class that sends metrics to the telemetry service on an interval.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

### I certified that

- [X] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [X] I discussed changes with code owners prior to submitting this pull request
- [X] I have not skipped any automated checks
- [X] All existing and new tests passed
- [X] I have updated the documentation accordingly

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/media/index.ts [hunk 1]
```diff
@@ -167,7 +170,14 @@ Media.createMediaConnection = (
       config.bundlePolicy = bundlePolicy;
     }
 
-    return new MultistreamRoapMediaConnection(config, debugId);
+    const metrics = new RtcMetrics(webex, debugId);
+
+    return new MultistreamRoapMediaConnection(
+      config,
+      debugId,
+      (data) => metrics.addMetrics(data),
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/media/index.ts:178]: so now we have the WCME doing getStats() every 5s and giving us result through this callback, but we also still have the SDK StatsAnalyzer calling getStats() every 5s to do its own analysis (to show logs when we don't send/receive packets etc and to send MQE). We should try to optimize it and avoid the duplicate calls to getStats(). I guess we should change the StatsAnalyzer to not call getStats() and instead use the stats results we get through this callback, do you agree?

BTW, this doesn't have to be fixed in this PR, but for sure needs to be addressed as part of this feature.

File: packages/@webex/plugin-meetings/test/unit/spec/rtcMetrics/index.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import RtcMetrics from '@webex/plugin-meetings/src/rtcMetrics';
+import MockWebex from '@webex/test-helper-mock-webex';
+import {assert} from '@webex/test-helper-chai';
+import sinon from 'sinon';
+
+const FAKE_METRICS_ITEM = {payload: ['fake-metrics']};
+
+describe.only('RtcMetrics', () => {
```
[the remark this answers is not in the document]
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/test/unit/spec/rtcMetrics/index.ts]: Removed.

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  /**
+   * Add metrics items to the metrics queue.
+   *
+   * @param {object} data - An object with a payload array of metrics items.
+   *
+   * @returns {void}
+   */
+  addMetrics(data) {
```
[the remark this answers is not in the document]
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/rtcMetrics/index.ts:50]: Not currently, will need to export from wcme and then internal-media-core.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 4]
```diff
@@ -4937,18 +4937,23 @@ export default class Meeting extends StatelessWebexPlugin {
...
-      mediaProperties: this.mediaProperties,
-      remoteQualityLevel: this.mediaProperties.remoteQualityLevel,
-      // @ts-ignore - config coming from registerPlugin
-      enableRtx: this.config.enableRtx,
-      // @ts-ignore - config coming from registerPlugin
-      enableExtmap: this.config.enableExtmap,
-      turnServerInfo,
-      bundlePolicy,
-    });
+  async createMediaConnection(turnServerInfo, bundlePolicy?: BundlePolicy) {
```
[the remark this answers is not in the document]
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: Not sure, changed back to private.

File: packages/@webex/plugin-meetings/src/media/index.ts [hunk 5]
```diff
@@ -167,7 +170,14 @@ Media.createMediaConnection = (
...
+      () => metrics.closeMetrics
```
[the remark this answers is not in the document]
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/media/index.ts]: Right, updated.

File: packages/@webex/plugin-meetings/src/media/index.ts (hunk 1 above)
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/media/index.ts:178]: No because the SDK StatsAnalyzer calls transceiver-specific stats which are formatted differently than these metrics getStats calls which are originally called on the PeerConnection. The stats in this PR are also formatted quite differently than the a regular `StatsReport` item, they are compressed and organized into a unified telemetry-friendly format.

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/rtcMetrics/index.ts:50]: ok, so not a must for the initial PR, but let's do that

File: packages/@webex/plugin-meetings/src/media/index.ts (hunk 1 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/media/index.ts:178]: ok, I'm sure we could refactor StatsAnalyzer and/or the rtcstats code so that we could have StatsAnalyzer working with this data and avoid doing 2x more calls to getStats() than necessary, but it's a bigger piece of work, and not sure if it's really necessary, so for now it's OK to keep it as it is and we'll see if calling getStats() so often actually causes any issues. Also when we have these dumps available to download, the StatsAnalyzer will become largely obsolete, so we could maybe just change it to do getStats() calls once per minute only so that it can still send MQEs. All of that for is something to look into later on, so for now I've created a separate jira for this: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-445740

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
+import uuid from 'uuid';
+
+const appId = 'FFB51ED5-4319-4C55-8303-B1F2FCCDE231';
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/rtcMetrics/index.ts:3]: i would move all the constants to a constants file in the rtcMetrics , can be done in next PR too

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   * @returns {void}
+   */
+  private sendMetrics() {
+    this.webex.request({
+      method: 'POST',
+      service: 'unifiedTelemetry',
+      resource: 'metric/v2',
+      headers: {
+        'Content-Type': 'application/json',
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/rtcMetrics/index.ts:83]: i dont think we need to pass the content-type , browser should auto add it based on body if I am not wrong

File: packages/@webex/plugin-meetings/src/media/index.ts [hunk 8]
```diff
@@ -173,7 +178,14 @@ Media.createMediaConnection = (
       config.bundlePolicy = bundlePolicy;
     }
 
-    return new MultistreamRoapMediaConnection(config, debugId);
+    const metrics = new RtcMetrics(webex, meetingId);
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/media/index.ts:181]: i would call this rtcMetrics to avoid confusion

File: packages/@webex/plugin-meetings/src/media/index.ts (hunk 1 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/media/index.ts:186]: AFAIK MQEs from web are ignored, so we could even remove it entirely...
I think the plan is for the telemetry service to compute the MQE format and send it over, so eventually we can just keep this rtcstats code.

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+   * @param {object} webex - The main `webex` object.
+   * @param {string} meetingId - The meeting id.
+   */
+  constructor(webex, meetingId) {
+    // `window` is used to prevent typescript from returning a NodeJS.Timer.
+    this.intervalId = window.setInterval(this.checkMetrics.bind(this), 30 * 1000);
+    this.meetingId = meetingId;
+    this.webex = webex;
+    this.correlationId = uuid.v4();
+    setTimeout(this.checkMetrics.bind(this), 5 * 1000);
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/rtcMetrics/index.ts:34]: can you add comments why is this needed

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 10]
```diff
@@ -37,9 +37,7 @@ import {
   RemoteTrackType,
   MediaType,
 } from '@webex/internal-media-core';
-import {
-  LocalTrackEvents,
-} from '@webex/media-helpers';
+import {LocalTrackEvents} from '@webex/media-helpers';
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:40]: please revert any formatting changes

Comment (torgeadelin) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:40]: not just in this file, but everywhere

File: packages/@webex/plugin-meetings/test/unit/spec/rtcMetrics/index.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  let webex: MockWebex;
+  let clock;
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers();
+    webex = new MockWebex();
+    metrics = new RtcMetrics(webex, 'mock-meeting-id');
+  });
+
+  it('should send metrics requests', () => {
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/test/unit/spec/rtcMetrics/index.ts:19]: you should test the other methods too

**Review by torgeadelin — CHANGES_REQUESTED** [review]

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts (hunk 6 above)
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/rtcMetrics/index.ts]: Added constants file.

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts (hunk 7 above)
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/rtcMetrics/index.ts]: Removed

File: packages/@webex/plugin-meetings/src/rtcMetrics/index.ts (hunk 9 above)
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/src/rtcMetrics/index.ts:34]: Added

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js (hunk 10 above)
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js]: Reverted

File: packages/@webex/plugin-meetings/test/unit/spec/rtcMetrics/index.ts (hunk 11 above)
Comment (mccarthytyler) [reply packages/@webex/plugin-meetings/test/unit/spec/rtcMetrics/index.ts]: Added more tests

**Review by torgeadelin — APPROVED** [review]

**Review by arun3528 — APPROVED** [review]
